pkg/cloud/openstack/clients/machineservice.go

```diff
@@ -162,7 +167,24 @@ func NewInstanceServiceFromCloud(cloud clientconfig.Cloud) (*InstanceService, er
 
 	opts.AllowReauth = true
 
-	provider, err := openstack.AuthenticatedClient(*opts)
+	provider, err := openstack.NewClient(opts.IdentityEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	cloudFromYaml, err := clientconfig.GetCloudFromYAML(clientOpts)
+	if cloudFromYaml.CACertFile != "" {
+		roots, err := certutil.NewPool(caFile)
```

Comment:
I am thinking about the mounted caFile here. In theory, the entire code allows us to use a separate `clouds.yaml` per machine. But it is only possible to pass in a caFile for all machines.

How about extending the `cloud-config` secret with another field for the contents of the caFile.... and create a separate `x509.CertPool` from this secret?

I think it would be ok to extend the `cloud-config` secret, because it relates directly to the `clouds.yaml`, rather than having a separate secret.

Reply:
@hidekazuna @gyliu513 provided this to me 
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/compare/master...cdc1807:cacert-auth

I think we can try to refer to this one, at least it works for me now on a https openstack and I think we can use secret to replace the CACERT in the commit diff

Reply:
@jichenjc Thanks advice.
I updated like this: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/672c5affd15e14b1fb829e620087a8fe8c5c20de#diff-6f1595bb05179a80513b3baba49fbaf1R58
I noticed I need to update getNetworkClient function in pkg/cloud/openstack/cluster/actuator.go.
But I wonder how  to fix this function as well. Could you advice?

Reply:
I might need more time to figure out but FYI https://github.com/kubernetes/cloud-provider-openstack has a set of https handling including network, maybe we can refer to their implementation @hidekazuna

Reply:
@chrigl @jichenjc getNetworkClient is the method of Actuator in cluster package and it's augment is *clusterv1.Cluster. We can not secrets from Actuator nor clusterv1.Cluster. I thinks this is from design issue.
Since we have to create a CA file for using OpenStack Cloud provider anyway, can't we make a compromise to use CA file?

Reply:
I am not aware clusterv1 can't use secrets from Actuator, ... but conceptually yes, if you want to sync cluster to create network, apparently we need CA in order to connect to openstack, so I think it's mandatory to use CA in actuator, can we define in openstackclusterspec for this ? @hidekazuna

Reply:
@jichenjc Thanks, now I am trying to update.

Reply:
@jichenjc I updated by https://github.com/kubernetes-sigs/cluster-api-provider-openstack/commit/0bf8a1bf20fa24ef22c193d48b2f5bf292527979